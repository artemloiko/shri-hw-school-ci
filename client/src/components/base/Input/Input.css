.input {
  width: fit-content;
  display: flex;
  justify-content: space-between;
  position: relative;
  --clear-button-width: calc(16px + (var(--space-input-h) * 2));
  --clear-button-height: calc(16px + (var(--space-input-top) + var(--space-input-bottom)));
}

.input_fullwidth {
  width: 100%;
}
.input_counter {
  width: 52px;
}
.input_counter .input__text {
  text-align: right;
}

.input__text {
  width: 100%;
  padding: var(--space-input-top) var(--space-input-h) var(--space-input-bottom);
  border: 2px solid var(--color-border-input);
  border-radius: var(--border-radius-m);

  font-size: var(--typo-size-s);
  line-height: 1.1539;

  color: var(--color-text);
  background-color: var(--color-bg-input);
}

.input__text:focus {
  outline: none;
  border-color: var(--color-border-input-focused);
}

.input__text::placeholder {
  color: var(--color-text-input-placeholder);
}

.input__clear-btn {
  display: none;
  position: absolute;
  width: var(--clear-button-width);
  height: var(--clear-button-height);
  top: 2px;
  right: 2px;

  cursor: pointer;
  background-size: 16px 16px;
  background-repeat: no-repeat;
  background-position: center;
  background-image: url('../../../images/icon/clear.svg');
}

.input_clear .input__text {
  padding-right: var(--clear-button-width);
}

.input_clear .input__clear-btn {
  display: block;
}

.input__text:placeholder-shown ~ .input__clear-btn {
  display: none;
}
