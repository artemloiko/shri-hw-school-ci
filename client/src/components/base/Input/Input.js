import React from 'react';
import PropTypes from 'prop-types';
import { cn } from 'utils/bem-cn';

import './Input.css';

function Input(props) {
  return (
    <div className={cn('input', props)}>
      <input {...props} className="input__text" />
      {props.mods.clear && <div className="input__clear-btn" tabIndex="0"></div>}
    </div>
  );
}

Input.propTypes = {
  type: PropTypes.string,
  id: PropTypes.string,
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  defaultValue: PropTypes.string,
  inputMode: PropTypes.string,

  className: PropTypes.string,
  mods: PropTypes.shape({
    fullwidth: PropTypes.bool,
    clear: PropTypes.bool,
    counter: PropTypes.bool,
  }),
  mix: PropTypes.arrayOf(PropTypes.string),
};

export default Input;
