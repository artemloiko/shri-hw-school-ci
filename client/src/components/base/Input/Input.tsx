import React, { useCallback, useRef } from 'react';
import { cn, CNProps } from '../../../utils/bem-cn';

import './Input.css';

type InputMods = {
  fullwidth?: boolean;
  clear?: boolean;
  counter?: boolean;
};

type Props = {
  setFieldValue: (name: string, value: string | number) => void;
};

type InputProps = Props & CNProps<InputMods> & React.InputHTMLAttributes<HTMLInputElement>;

const Input: React.FC<InputProps> = (props) => {
  const { setFieldValue, mods = {}, className, mix, name, onBlur, ...inputProps } = props;
  const inputRef = useRef<HTMLInputElement>(null);

  const handleClear = useCallback(
    (e) => {
      if (e.type === 'keydown' && e.key !== ' ' && e.key !== 'Enter') {
        return;
      }
      e.preventDefault();
      if (name) {
        setFieldValue(name, '');
        inputRef?.current?.focus();
      }
    },
    [setFieldValue, name],
  );

  const handleBlur = useCallback(
    (e: React.FocusEvent<HTMLInputElement>) => {
      if (e.currentTarget.value && name) {
        setFieldValue(name, e.currentTarget.value.trim());
      }
      if (onBlur) onBlur(e);
    },
    [onBlur, setFieldValue, name],
  );

  return (
    <div className={cn('input', { className, mods, mix })}>
      <input {...inputProps} className="input__text" ref={inputRef} onBlur={handleBlur} />
      {mods.clear && (
        <div
          className="input__clear-btn"
          tabIndex={0}
          onClick={handleClear}
          onKeyDown={handleClear}
          role="button"
          aria-label="Clear input"
        ></div>
      )}
    </div>
  );
};

export default Input;
