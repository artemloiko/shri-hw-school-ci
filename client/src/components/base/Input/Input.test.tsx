import React from 'react';
import { create } from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';

import Input from './Input';

test('Input with all possible props matches the snapshot', () => {
  const tree = create(
    <Input
      id="commitHash"
      name="commitHash"
      placeholder="Commit hash"
      type="text"
      inputMode="text"
      required
      pattern="\w{7,}"
      title="Hash 7 letters/numbers"
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      onChange={(): void => {}}
      onBlur={(): void => {}}
      setFieldValue={(): void => {}}
      className="block"
      mods={{ clear: true, fullwidth: true }}
      mix={['block']}
    />,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

test('Input is trimming value on blur via setFieldValue call', () => {
  const setFieldValueMock = jest.fn();
  const { getByPlaceholderText } = render(
    <Input
      id="commitHash"
      name="commitHash"
      placeholder="Commit hash"
      onChange={(): void => {}}
      onBlur={(): void => {}}
      setFieldValue={setFieldValueMock}
      mods={{ clear: true }}
      data-testid="input"
    />,
  );

  fireEvent.blur(getByPlaceholderText('Commit hash'), { target: { value: '  trim  ' } });

  expect(setFieldValueMock).toHaveBeenCalledWith('commitHash', 'trim');
});

describe('Input clear button', () => {
  test('is avaible for focus', () => {
    const { getByRole } = render(
      <Input
        onChange={(): void => {}}
        onBlur={(): void => {}}
        setFieldValue={(): void => {}}
        mods={{ clear: true }}
      />,
    );
    const clearButton = getByRole('button');

    expect(clearButton).toHaveAttribute('tabindex', '0');
  });

  test('is clearing value via setFieldValue(name, value)', () => {
    const setFieldValueMock = jest.fn();
    const { getByRole } = render(
      <Input
        name="input"
        onChange={(): void => {}}
        onBlur={(): void => {}}
        setFieldValue={setFieldValueMock}
        mods={{ clear: true }}
      />,
    );
    const clearButton = getByRole('button');

    fireEvent.click(clearButton);

    expect(setFieldValueMock).toHaveBeenCalledWith('input', '');
  });

  test('is clearing on events: [click, keyDown.enter, keyDown.space]', () => {
    const setFieldValueMock = jest.fn();
    const { getByRole } = render(
      <Input
        name="input"
        onChange={(): void => {}}
        onBlur={(): void => {}}
        setFieldValue={setFieldValueMock}
        mods={{ clear: true }}
      />,
    );
    const clearButton = getByRole('button');

    fireEvent.click(clearButton);
    fireEvent.keyDown(clearButton, { key: 'Enter' });
    fireEvent.keyDown(clearButton, { key: ' ' });

    expect(setFieldValueMock).toHaveBeenCalledTimes(3);
  });

  test('is focusing input after clear', () => {
    const { getByRole, getByPlaceholderText } = render(
      <Input
        placeholder="input"
        onChange={(): void => {}}
        onBlur={(): void => {}}
        setFieldValue={(): void => {}}
        mods={{ clear: true }}
        name="input"
      />,
    );
    const input = getByPlaceholderText('input');
    const clearButton = getByRole('button');

    fireEvent.click(clearButton);

    expect(input).toHaveFocus();
  });
});
