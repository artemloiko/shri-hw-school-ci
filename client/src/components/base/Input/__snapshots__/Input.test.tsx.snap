// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input with all possible props matches the snapshot 1`] = `
<div
  className="block input input_clear input_fullwidth block__input"
>
  <input
    className="input__text"
    id="commitHash"
    inputMode="text"
    onBlur={[Function]}
    onChange={[Function]}
    pattern="\\\\w{7,}"
    placeholder="Commit hash"
    required={true}
    title="Hash 7 letters/numbers"
    type="text"
  />
  <div
    aria-label="Clear input"
    className="input__clear-btn"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex={0}
  />
</div>
`;
